### PR TITLE
Kpoints line mode for VASP

### PR DESCRIPTION
**Pull request by FeLoch**
added line mode and related stuff:

- line mode works
- high-symmetry points as structure property

### CODE REVIEW

File: .cache/v/cache/lastfailed [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (jan-janssen) [inline .cache/v/cache/lastfailed]: Please remove these files and add them to the `.gitiginore`

File: pyiron.egg-info/PKG-INFO [hunk 2]
```diff
@@ -0,0 +1,22 @@
+Metadata-Version: 1.1
```
Comment (jan-janssen) [inline pyiron.egg-info/PKG-INFO]: Please remove these files and add them to the `.gitiginore`

File: pyiron.egg-info/SOURCES.txt [hunk 3]
```diff
@@ -0,0 +1,121 @@
+MANIFEST.in
```
Comment (jan-janssen) [inline pyiron.egg-info/SOURCES.txt]: Please remove these files and add them to the `.gitiginore`

File: pyiron.egg-info/dependency_links.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+
```
Comment (jan-janssen) [inline pyiron.egg-info/dependency_links.txt]: Please remove these files and add them to the `.gitiginore`

File: pyiron.egg-info/requires.txt [hunk 5]
```diff
@@ -0,0 +1,18 @@
+ase
```
Comment (jan-janssen) [inline pyiron.egg-info/requires.txt]: Please remove these files and add them to the `.gitiginore`

File: pyiron.egg-info/top_level.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+pyiron
```
Comment (jan-janssen) [inline pyiron.egg-info/top_level.txt]: Please remove these files and add them to the `.gitiginore`

File: venv/Scripts/Activate.ps1 [hunk 7]
```diff
@@ -0,0 +1,51 @@
+function global:deactivate ([switch]$NonDestructive) {
```
Comment (jan-janssen) [inline venv/Scripts/Activate.ps1]: Please remove these files and add them to the `.gitiginore`

File: venv/Scripts/activate [hunk 8]
```diff
@@ -0,0 +1,76 @@
+# This file must be used with "source bin/activate" *from bash*
```
Comment (jan-janssen) [inline venv/Scripts/activate]: Please remove these files and add them to the `.gitiginore`

File: venv/Scripts/activate.bat [hunk 9]
```diff
@@ -0,0 +1,45 @@
+@echo off
```
Comment (jan-janssen) [inline venv/Scripts/activate.bat]: Please remove these files and add them to the `.gitiginore`

File: venv/Scripts/deactivate.bat [hunk 10]
```diff
@@ -0,0 +1,21 @@
+@echo off
```
Comment (jan-janssen) [inline venv/Scripts/deactivate.bat]: Please remove these files and add them to the `.gitiginore`

File: venv/pyvenv.cfg [hunk 11]
```diff
@@ -0,0 +1,3 @@
+home = C:\Users\Felix\Miniconda3
```
Comment (jan-janssen) [inline venv/pyvenv.cfg]: Please remove these files and add them to the `.gitiginore`

**Comment by FeLoch** [comment]
deleted! Sorry :D

**Comment by FeLoch** [comment]
done!

File: pyiron/base/generic/parameters.py [hunk 12]
```diff
@@ -471,9 +471,9 @@ def set_value(self, line, val):
             new_array.append(val)
             new_comments.append("")
             new_params.append("")
-            new_array = np.array(new_array)
-            new_comments = np.array(new_comments)
-            new_params = np.array(new_params)
+            new_array = np.array(new_array).tolist()
```
Comment (jan-janssen) [inline pyiron/base/generic/parameters.py]: Is this part related to the kpoints? Otherwise I would suggest we move this to a separate pull request, because I am not yet sure if it might break some functionality elsewhere.

Comment (sudarsan-surendralal) [reply pyiron/base/generic/parameters.py]: We can create a separate pull request. @FeLoch, shall I do this?

Comment (FeLoch) [reply pyiron/base/generic/parameters.py]: please, go ahead!

Comment (sudarsan-surendralal) [reply pyiron/base/generic/parameters.py]: Done here https://github.com/pyiron/pyiron/pull/446.

File: pyiron/atomistics/structure/atoms.py [hunk 13]
```diff
@@ -227,6 +228,7 @@ def __init__(
             else:
                 self.pbc = pbc
         self.set_initial_magnetic_moments(magmoms)
+        self._high_symmetry_points = high_symmetry_points
```
Comment (sudarsan-surendralal) [inline pyiron/atomistics/structure/atoms.py]: I think it might be better to do something like

```
if high_symmetry_points is not None:
   self.set_high_symmetry_points(high_symmetry_points)
```
This way, it will nicely raise an error if this parameter is not a `dict`

File: pyiron/atomistics/structure/atoms.py [hunk 14]
```diff
@@ -306,6 +308,24 @@ def elements(self):
         """
         return np.array([self.species[el] for el in self.indices])
 
+    def get_high_symmetry_points(self):
```
Comment (sudarsan-surendralal) [inline pyiron/atomistics/structure/atoms.py]: Just a small description in docstrings will be good

File: pyiron/vasp/base.py [hunk 15]
```diff
@@ -1029,8 +1031,10 @@ def _set_kpoints(
             reciprocal (bool): Tells if the supplied values are in reciprocal (direct) or cartesian coordinates (in
             reciprocal space)
             kmesh_density (float): Value of the required density
+            n_trace (int): Number of points per trace part for line mode
+            trace (list): ordered list of high symmetry points for line mode
         """
-
+        self.structure.get_high_symmetry_points()
```
Comment (sudarsan-surendralal) [inline pyiron/vasp/base.py]: This line does nothing I guess

File: pyiron/vasp/base.py [hunk 16]
```diff
@@ -1662,11 +1705,12 @@ def from_hdf(self, hdf):
             self.incar.from_hdf(hdf5_input)
             self.kpoints.from_hdf(hdf5_input)
             self.potcar.from_hdf(hdf5_input)
+
+            self._eddrmm = "ignore"
```
Comment (sudarsan-surendralal) [inline pyiron/vasp/base.py]: Isn't the default already specified? Why do you need this?

File: pyiron/vasp/base.py [hunk 17]
```diff
@@ -1068,6 +1084,28 @@ def _set_kpoints(
...
+        gets the trace for k-points line mode in a VASP readable form.
+
+        Args:
+            trace (list): ordered list of names for k-points trace
+
+        Returns:
+            list: trace points coordinates for VASP
+        """
+        for t in trace:
+            if not t in self.structure.get_high_symmetry_points().keys():
```
Comment (sudarsan-surendralal) [inline pyiron/vasp/base.py]: pythonic ways is to write  `if t not in self.structure.get_high_symmetry_points().keys()`

File: pyiron/base/generic/parameters.py (hunk 12 above)
Comment (jan-janssen) [reply pyiron/base/generic/parameters.py]: @FeLoch Is this change necessary for your other changes to work?

File: pyiron/vasp/base.py (hunk 15 above)
Comment (FeLoch) [reply pyiron/vasp/base.py]: is already deleted

File: pyiron/vasp/base.py (hunk 16 above)
Comment (FeLoch) [reply pyiron/vasp/base.py]: we need this for older job without the tag. The default for new jobs is 'not_converged'

File: pyiron/base/generic/parameters.py (hunk 12 above)
Comment (FeLoch) [reply pyiron/base/generic/parameters.py]: yes it is. Without these changes the file would be converted in an array, if you increased the number of lines compared to the default file. Therefore, those changes are only necessary for line mode calculations

File: pyiron/vasp/base.py (hunk 17 above)
Comment (FeLoch) [reply pyiron/vasp/base.py]: done

File: pyiron/atomistics/structure/atoms.py (hunk 14 above)
Comment (FeLoch) [reply pyiron/atomistics/structure/atoms.py]: done

File: pyiron/atomistics/structure/atoms.py (hunk 13 above)
Comment (FeLoch) [reply pyiron/atomistics/structure/atoms.py]: done

File: pyiron/vasp/base.py [hunk 18]
```diff
@@ -2135,15 +2178,52 @@ def __init__(self, input_file_name=None, table_name="kpoints"):
...
             size_of_mesh (list/numpy.ndarray): List of size 1x3 specifying the required mesh size
             shift (list): List of size 1x3 specifying the user defined shift from the Gamma point
+            n_trace (int): Number of points per trace for line mode
+            trace_coord (list): coordinates of k-points trace in VASP KPOINTS format
         """
+        if n_trace is not None:
+            if self._trace is None or trace_coord is None:
+                raise ValueError("trace have to be defined")
+
+            self.set_value(line=1, val=n_trace)
```
Comment (jan-janssen) [inline pyiron/vasp/base.py]: The Kpoints file is the only one that uses `set_value()` instead of `self[1] = n_trace` - @sudarsan-surendralal Is there a specific reason for this?

Comment (sudarsan-surendralal) [reply pyiron/vasp/base.py]: This would not work since in this case, the line number does not exist till now
```
from pyiron import Project
pr = Project("setval_demo")
job = pr.create_job(pr.job_type.Vasp, "setval_demo")
print(job.input.kpoints._dataset)
job.input.kpoints[5] = "5 5 5"

>> ---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-5-2614e019c818> in <module>
      3 job = pr.create_job(pr.job_type.Vasp, "setval_demo")
      4 print(job.input.kpoints._dataset)
----> 5 job.input.kpoints[5] = "5 5 5"

/cmmc/u/chandu/programs/pyiron_mpie/pyiron/pyiron/base/generic/parameters.py in __setitem__(self, key, value)
    640             if self.read_only and self._dataset["Value"][key] != value:
    641                 self._read_only_error()
--> 642             self._dataset["Value"][key] = value
    643         else:
    644             self.set(**{key: value})

IndexError: list assignment index out of range
```
Therefore, we have to stick to `set_value`

File: pyiron/atomistics/structure/atoms.py [hunk 19]
```diff
@@ -227,6 +228,10 @@ def __init__(
             else:
                 self.pbc = pbc
         self.set_initial_magnetic_moments(magmoms)
+        self._high_symmetry_points = None
+        if high_symmetry_points is not None:
+            self.set_high_symmetry_points(high_symmetry_points)
+
```
Comment (jan-janssen) [inline pyiron/atomistics/structure/atoms.py]: Please remove this empty line
